### DISCUSSION TITLE
Implement ordered traversal and traversal depth

### DISCUSSION THREAD

**Discussion by oknozor** [GitoxideLabs/gitoxide#617] [Ideas]
### Summary 💡

As mentioned here the git-traverse crate only support non-ordered breadthfirst traversal. 
https://github.com/Byron/gitoxide/blob/ff9e1571b558475e727dc6ba11dab24ef15fb6f4/git-repository/src/object/tree/traverse.rs#L47-L48

### Motivation 🔦

I am building a git platform using gitoxide. 

The tree explorer you can see in the screenshot below use the current breadth-first search traversal. 

The current implementation works fine but needs to traverse the entire tree from root each time we load a page. An alternative would be to load the tree in order with a given depth to get only what needs to be rendered.

## Example
```
. // root
├── one
│   ├── two // depth 1
│   │   └── blob // depth 2
│   └── tree // depth 1
│       ├── blob // depth 2
│       ├── blob // .. 
│       ├── blob // ..
```

And this is the real life use case : 
![screenshot-2022-11-20T21:58:21](https://user-images.githubusercontent.com/24563836/202925833-b8ad85f6-c9bd-48a7-b1e7-4d806e3629e9.png)

I have no idea if this is doable, I will try to implement it if you are ok with the idea. 
Otherwise I will just cache everything :)

**Comment by Byron** [comment]
I think all information that's needed to make traversal efficient is in the `Delegate` implemention. It knows which level entries are on, and it can cancel the iteration.

[Here information can be tracked](https://github.com/Byron/gitoxide/blob/34e899ae0e4a23bba6dc36dfd7429c0d572736bd/git-traverse/src/tree/breadthfirst.rs#L67-L94) pertaining the current level of iteration. Once [this method](https://github.com/Byron/gitoxide/blob/34e899ae0e4a23bba6dc36dfd7429c0d572736bd/git-traverse/src/tree/breadthfirst.rs#L98) is called we know that we are about to enter another level.

However, for this problem, it appears all you'd want is to iterate all entries of a single level, best done with [`.iter()`](https://github.com/Byron/gitoxide/blob/6dc3ec1936b8c74e162e95a5aa1ff0a0d13e6fc8/git-repository/src/object/tree/iter.rs#L48-L49).

I hope that helps.
